relayer/src/chain/cosmos.rs

```diff
@@ -336,45 +336,58 @@ impl CosmosSdkChain {
-    /// if the batch is split in two TX-es, the second one will fail the simulation in `deliverTx` check
+    ///
+    /// If the batch is split in two TX-es, the second one will fail the simulation in `deliverTx` check.
     /// In this case we use the `default_gas` param.
     fn estimate_gas(&self, tx: Tx) -> Result<u64, Error> {
-        let estimated_gas = self.send_tx_simulate(tx).map(|sr| sr.gas_info);
+        let simulated_gas = self.send_tx_simulate(tx).map(|sr| sr.gas_info);
 
-        if let Ok(ref gas) = estimated_gas {
-            debug!(
-                "[{}] send_tx: tx simulation successful, simulated gas: {:?}",
-                self.id(),
-                gas
-            );
-        }
+        match simulated_gas {
+            Ok(Some(gas_info)) => {
+                debug!(
+                    "[{}] estimate_gas: tx simulation successful, gas amount used: {:?}",
+                    self.id(),
+                    gas_info.gas_used
+                );
 
-        let estimated_gas = estimated_gas.map_or_else(
-            |e| {
-                error!(
-                    "[{}] send_tx: failed to estimate gas, falling back on default gas, error: {}",
+                Ok(gas_info.gas_used)
+            }
+
+            Ok(None) => {
+                warn!(
+                    "[{}] estimate_gas: tx simulation successful but no gas amount used was returned, falling back on default gas: {}",
+                    self.id(),
+                    self.default_gas()
+                );
+
+                Ok(self.default_gas())
+            }
+
+            // If there is a chance that the tx will be accepted once actually submitted, we fall
+            // back on the max gas and will attempt to send it anyway.
+            // See `can_recover_from_simulation_failure` for more info.
+            Err(e) if can_recover_from_simulation_failure(&e) => {
+                warn!(
+                    "[{}] estimate_gas: failed to simulate tx, falling back on default gas because the error is potentially recoverable: {}",
                     self.id(),
                     e.detail()
                 );
 
-                self.default_gas()
-            },
-            |gas_info| gas_info.map_or(self.default_gas(), |g| g.gas_used),
-        );
+                Ok(self.default_gas())
+            }
 
-        if estimated_gas > self.max_gas() {
-            debug!(estimated = ?estimated_gas, max = ?self.max_gas(), "[{}] send_tx: estimated gas is higher than max gas", self.id());
+            Err(e) => {
+                error!(
+                    "[{}] estimate_gas: failed to simulate tx with non-recoverable error: {}",
+                    self.id(),
+                    e.detail()
+                );
 
-            return Err(Error::tx_simulate_gas_estimate_exceeded(
-                self.id().clone(),
-                estimated_gas,
-                self.max_gas(),
-            ));
+                Err(e)
+            }
```

Comment:
We should still check that the estimated gas is smaller than the configured max gas, no?

Reply:
https://github.com/informalsystems/ibc-rs/pull/1542/commits/fedd6c6af9c51f84c7d55075073a3024cecf77ae & https://github.com/informalsystems/ibc-rs/pull/1542/commits/fedd6c6af9c51f84c7d55075073a3024cecf77ae